lib/keystore.cc

```diff
@@ -82,55 +130,113 @@ rpmRC keystore_fs::delete_key(rpmtxn txn, rpmPubkey key)
 rpmRC keystore_fs::import_key(rpmtxn txn, rpmPubkey key, int replace, rpmFlags flags)
 {
     rpmRC rc = RPMRC_FAIL;
-    const char *fp = rpmPubkeyFingerprintAsHex(key);
-    char *keyfmt = rstrscat(NULL, "gpg-pubkey-", fp, ".key", NULL);
-    char *keyval = rpmPubkeyArmorWrap(key);
-    char *path = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", keyfmt);
-    char *tmppath = NULL;
+    string fp = rpmPubkeyFingerprintAsHex(key);
+    string keyfmt = "gpg-pubkey-" + fp + ".key";
+    char *dir = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", NULL);
+    string dirstr = dir;
 
-    if (replace) {
-	rasprintf(&tmppath, "%s.new", path);
-	unlink(tmppath);
+    rc = write_key_to_disk(key, dirstr, keyfmt, replace, flags);
+
+    if (!rc && replace) {
+	/* find and delete the old pubkey entry */
+	if (delete_key(txn, fp, keyfmt) == RPMRC_NOTFOUND) {
+	    /* make sure an old, short keyid version gets removed */
+	    delete_key(txn, fp.substr(32), keyfmt);
+	}
     }
 
-    if (rpmMkdirs(rpmtxnRootDir(txn), "%{_keyringpath}")) {
+    free(dir);
+    return rc;
+}
+
+/*****************************************************************************/
+
+static rpmRC acquire_write_lock(rpmtxn txn)
```

Comment:
The txn handle *is the lock*, that's the whole point of those things. There's no other locking needed.

Reply:
I know. But the standard requires a lockfile on disk as part of the openpgp.cert.d format. If we want other tools to be able to read it we should try to adhere to that.

Reply:
I don't see concurrent access with other tools being something we need to support. Poking at rpm directories with external tools is more like a developer convenience thing rather than something we'd be suggesting to users. But, if you want to support locking then the actual lock seems to be missing, https://www.ietf.org/archive/id/draft-nwjw-openpgp-cert-d-00.html#platform-specifics says it should use flock().

Reply:
...ie it's not the existence of 'writelock' file that is the lock, but a flock() on that file. Mere file existence would be broken lock mechanism because if the process dies in the middle it de-facto leaves a stale lock behind, whereas locks on the file are cleaned up by the OS.

Reply:
*urgs* should have read the last sentence, too...
If I read that correctly you are not actually required to delete `writelock` the file again. Just to be fine if it doesn't exist. @nwalfield any opinion on that?

Anyway implemented the flock and adjusted the test case.

Reply:
> If I read that correctly you are not actually required to delete writelock the file again. Just to be fine if it doesn't exist.

Yup, that's how such lockfiles commonly work, including our own transaction lock.